### PR TITLE
W2v test loss

### PR DESCRIPTION
**Pull request by evsmithx**

### CODE REVIEW

File: libs/ml/examples/model_saver.hpp [hunk 1]
```diff
@@ -29,28 +32,175 @@ namespace ml {
...
+}
+
+/**
+ * Saves the saveparams of a graph to a file location specified by user.
+ * Is quicker than SaveModel for large graphs.
+ * @param g the graph to save
+ * @param save_location a string specifying save location
+ */
+template <typename GraphType>
+void SaveLargeModel(GraphType &g, std::string const &save_location)
```
Comment (baykaner) [inline libs/ml/examples/model_saver.hpp]: SaveModel and SaveLargeModel seems like a fair bit of duplicated code.
How about just having SaveLargeModel (but rename it to SaveModel).
Or if you really need both functionality, you can give it a boolean flag defaulting to large model?

File: libs/ml/examples/model_saver.hpp [hunk 2]
```diff
@@ -29,28 +32,175 @@ namespace ml {
...
+void SaveObject(MLType &g, std::string const &save_location)
+{
+  std::cout << "Starting object serialization" << std::endl;
+
   fetch::serializers::MsgPackSerializer serializer;
-  serializer << g.StateDict();
+
+  serializer << g;
 
   std::fstream file(save_location, std::fstream::out);  // fba = FetchByteArray
```
Comment (baykaner) [inline libs/ml/examples/model_saver.hpp]: is this comment useful?

File: libs/ml/examples/model_saver.hpp [hunk 3]
```diff
@@ -29,28 +32,175 @@ namespace ml {
...
+  return graph_ptr;
+}
+
+/**
+ * Saves an object to a file location specified by user.
+ * @param g the graph to save
+ * @param save_location a string specifying save location
+ */
+template <typename MLType>
+void SaveObject(MLType &g, std::string const &save_location)
```
Comment (baykaner) [inline libs/ml/examples/model_saver.hpp]: SaveObject and SaveModel and SaveLargeModel all look suspiciously similar. Is there some duplication here?

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 4]
```diff
@@ -59,76 +62,10 @@ std::pair<std::string, std::string> Model(fetch::ml::Graph<TensorType> &g, SizeT
...
-
-  for (std::size_t l = 0; l < output.size(); ++l)
-  {
-    std::cout << "rank: " << l << ", "
-              << "distance, " << output.at(l).second << ": " << dl.WordFromIndex(output.at(l).first)
-              << std::endl;
-  }
-}
-
 void TestEmbeddings(Graph<TensorType> const &g, std::string const &skip_gram_name,
```
Comment (baykaner) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp:65]: I wonder if testEmbeddings should live in the dataloader? that seems more like a general utility function?

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 5]
```diff
@@ -157,46 +90,49 @@ std::string ReadFile(std::string const &path)
 
 struct TrainingParams
 {
-  // TODO (#1585) something is broken here. if u set max_word_count to sth smaller like 10000, there
-  // would be an error at the end of the sentence
+  // TODO (#1585) something is broken here. If you set max_word_count to something smaller like
```
Comment (baykaner) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp]: I think this TODO is fixed right? Please check and remove as necessary

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 6]
```diff
@@ -157,46 +90,49 @@ std::string ReadFile(std::string const &path)
...
+  // 10000, there would be an error at the end of the sentence
   SizeType max_word_count = fetch::math::numeric_max<SizeType>();  // maximum number to be trained
-  SizeType negative_sample_size = 5;     // number of negative sample per word-context pair
-  SizeType window_size          = 5;     // window size for context sampling
-  DataType freq_thresh          = 1e-3;  // frequency threshold for subsampling
-  SizeType min_count            = 5;     // infrequent word removal threshold
+  SizeType negative_sample_size = 5;      // number of negative sample per word-context pair
+  SizeType window_size          = 2;      // window size for context sampling
```
Comment (baykaner) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp:98]: is this usual? I don't remember what Mikolov et al. set for window size but I thought it was a max of 4 or 5? not 2?

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 7]
```diff
@@ -157,46 +90,49 @@ std::string ReadFile(std::string const &path)
...
+  DataType freq_thresh          = 1e-3f;  // frequency threshold for subsampling
+  SizeType min_count            = 100;    // infrequent word removal threshold
```
Comment (baykaner) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp:100]: is this not quite high?
Unless a word occurred 100 times we ditch it?

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp [hunk 8]
```diff
@@ -157,46 +90,49 @@ std::string ReadFile(std::string const &path)
...
 
-  SizeType batch_size      = 100000;  // training data batch size
-  SizeType embedding_size  = 100;     // dimension of embedding vec
+  SizeType batch_size      = 10000;  // training data batch size
+  SizeType embedding_size  = 500;    // dimension of embedding vec
```
Comment (baykaner) [inline libs/ml/examples/word2vec_SGNS_dataloader/main.cpp:103]: pretty large default embeddings!
if these hyperparameters are to match an implementation in a paper somewhere please add a comment above them explaining this and giving a link to that paper

File: libs/ml/examples/word2vec_graph_evaluation/main.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+//
+//------------------------------------------------------------------------------
+
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp"
+#include "ml/layers/skip_gram.hpp"
+#include "word2vec_utilities.hpp"
+
+#include <model_saver.hpp>
```
Comment (baykaner) [inline libs/ml/examples/word2vec_graph_evaluation/main.cpp]: "" not <>

File: libs/ml/examples/word2vec_utilities.hpp [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+{
+  using SizeType = typename TensorType::SizeType;
+  using DataType = typename TensorType::Type;
+
+  // get word vectors for word_ids
+  TensorType word1_vec = embeddings.Slice(word1, 1).Copy();
+  TensorType word2_vec = embeddings.Slice(word2, 1).Copy();
+  TensorType word3_vec = embeddings.Slice(word3, 1).Copy();
+
+  //  word1_vec /= fetch::math::L2Norm(word1_vec);
```
Comment (baykaner) [inline libs/ml/examples/word2vec_utilities.hpp]: you don't normalise these anymore? I thought that was important (to ensure you're really just comparing the angle between entities in embeddings space, rather than absolute distance, which may be less relevant)?

In any case, left in commented out code

File: libs/ml/examples/word2vec_utilities.hpp [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+  //  word2_vec /= fetch::math::L2Norm(word2_vec);
+  //  word3_vec /= fetch::math::L2Norm(word3_vec);
+
+  TensorType                                 word4_vec = word2_vec - word1_vec + word3_vec;
+  std::vector<std::pair<SizeType, DataType>> output;
```
Comment (baykaner) [inline libs/ml/examples/word2vec_utilities.hpp]: lines 53 & 54 can be one line

File: libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp [hunk 12]
```diff
@@ -417,11 +417,7 @@ void GraphW2VLoader<T>::BufferNextSamples()
   // set up a counter to add samples to buffer
   SizeType counter = 0;
 
-  // reset all three buffers
   input_words_.Fill(cur_word_id);
-  labels_.Fill(BufferPositionUnused);
```
Comment (baykaner) [inline libs/ml/include/ml/dataloaders/word2vec_loaders/sgns_w2v_dataloader.hpp:423]: as discussed, you probably shouldn't delete these

**Review by baykaner — CHANGES_REQUESTED** [review]

File: libs/ml/examples/word2vec_SGNS_dataloader/main.cpp (hunk 4 above)
Comment (evsmithx) [reply libs/ml/examples/word2vec_SGNS_dataloader/main.cpp:65]: It's only used here - it's similar to one in the distributed example, but the distributed w2v client needs to lock the model and console so it's not the same. Maybe we'll move it if we reuse it somewhere?

**Review by baykaner — APPROVED** [review]
